### PR TITLE
sasarkar/qwen finetuning bucketing

### PR DESCRIPTION
**Pull request by ssarkar2**
# What does this PR do?
Add bucketing to Qwen training

earlier version: https://github.com/huggingface/optimum-habana/pull/1128

results:
```
time python3 sft.py  --model_name_or_path /root/sasarkar/Qwen2_7B/ --dataset_name "philschmid/dolly-15k-oai-style"  --streaming False  --bf16 True  --subset ''  --output_dir ./model_qwen  --num_train_epochs 1  --per_device_train_batch_size 4  --evaluation_strategy "no"  --save_strategy "no"  --learning_rate 3e-4  --warmup_ratio  0.03  --lr_scheduler_type "cosine"  --max_grad_norm  0.3  --logging_steps 1  --do_train  --do_eval  --use_habana  --use_lazy_mode  --throughput_warmup_steps 3  --lora_r 4  --lora_alpha=16  --lora_dropout=0.05  --lora_target_modules "q_proj" "v_proj" "k_proj" "o_proj"  --max_seq_length 512  --adam_epsilon 1e-08 --packing False --num_bucket 8
```

```
{'train_runtime': 1218.7741, 'train_samples_per_second': 12.185, 'train_steps_per_second': 3.046, 'train_loss': 1.5069085590719007, 'epoch': 1.0, 'memory_allocated (GB)': 31.57, 'max_memory_allocated (GB)': 39.44, 'total_memory_available (GB)': 94.62}
***** train metrics *****
  epoch                       =         1.0
  max_memory_allocated (GB)   =       39.44
  memory_allocated (GB)       =       31.57
  total_flos                  = 230347401GF
  total_memory_available (GB) =       94.62
  train_loss                  =      1.5069
  train_runtime               =  0:20:18.77
  train_samples_per_second    =      12.185
  train_steps_per_second      =       3.046
100%|██████████| 94/94 [00:22<00:00,  4.13it/s]
***** eval metrics *****
  epoch                       =        1.0
  eval_loss                   =     1.5511
  eval_runtime                = 0:00:22.83
  eval_samples                =        751
  eval_samples_per_second     =     32.141
  eval_steps_per_second       =      4.023
  max_memory_allocated (GB)   =      75.66
  memory_allocated (GB)       =      31.57
  perplexity                  =     4.7165
  total_memory_available (GB) =      94.62

real    21m31.841s
user    35m22.686s
sys     42m57.211s

```

Numbers without this branch:
```
{'train_runtime': 8899.1643, 'train_samples_per_second': 1.611, 'train_steps_per_second': 0.403, 'train_loss': 1.5117962875673847, 'epoch': 1.0, 'memory_allocate
d (GB)': 27.85, 'max_memory_allocated (GB)': 34.25, 'total_memory_available (GB)': 94.62}
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████
██████████| 3565/3565 [2:28:19<00:00,  2.50s/it]
***** train metrics *****
  epoch                       =         1.0
  max_memory_allocated (GB)   =       34.25
  memory_allocated (GB)       =       27.85
  total_flos                  = 184625734GF
  total_memory_available (GB) =       94.62
  train_loss                  =      1.5118
  train_runtime               =  2:28:19.16
  train_samples_per_second    =       1.611
  train_steps_per_second      =       0.403
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████
████████████████| 94/94 [00:49<00:00,  1.89it/s]
***** eval metrics *****
  epoch                       =        1.0
  eval_loss                   =     1.5515
  eval_runtime                = 0:00:49.63
  eval_samples                =        751
  eval_samples_per_second     =     14.713
  eval_steps_per_second       =      1.842
  max_memory_allocated (GB)   =      49.32
  memory_allocated (GB)       =      27.85
  perplexity                  =     4.7184
  total_memory_available (GB) =      94.62

real    150m3.097s
user    308m25.621s
sys     106m33.356s

```

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by regisss** [comment]
@ssarkar2 This will only work for Qwen? Or is it also appliable with other models?

**Comment by ssarkar2** [comment]
> @ssarkar2 This will only work for Qwen? Or is it also appliable with other models?

Haven't tried other models, but should work for others, since change is in SFT trainer

**Comment by ssarkar2** [comment]
8x cmd:

```
time DEEPSPEED_HPU_ZERO3_SYNC_MARK_STEP_REQUIRED=1 python3 ../ga
udi_spawn.py --world_size 2 --use_deepspeed sft.py  --model_name_or_path /mnt/weka/data/Qwen/Qwen2-72B/ --dataset_name "philschmid/dolly-15k-oai-style"  --stream
ing False  --bf16 True  --subset "''"  --output_dir ./model_qwen_70b  --num_train_epochs 1  --per_device_train_batch_size 16 --evaluation_strategy "no"  --save_
strategy "no"  --learning_rate 3e-4  --warmup_ratio  0.03  --lr_scheduler_type "cosine"  --max_grad_norm  0.3  --logging_steps 1  --do_train  --do_eval  --use_ha
bana  --use_lazy_mode  --throughput_warmup_steps 3  --lora_r 4  --lora_alpha=16  --lora_dropout=0.05  --lora_target_modules "q_proj" "v_proj" "k_proj" "o_proj"
--max_seq_length 512  --adam_epsilon 1e-08 --packing False --num_bucket 8 --max_steps 100
```

**Comment by ssarkar2** [comment]
TODO:
make qwen2-70b run using lora:

The command i was using earlier didnt have " --deepspeed " maybe that's why?

```
python3 ../gaudi_spawn.py --use_deepspeed --world_size 8 sft.py \
    --model_name_or_path "/root//mnt/weka/data/Qwen/Qwen2-72B/" \
    --dataset_path "/root/litang/QWEN/qwen2_finetune_bench_scripts/belle_chat_ramdon_10k.json" \
    --bf16 True \
    --output_dir ./model_lora_qwen \
    --num_train_epochs 5 \
    --per_device_train_batch_size 10 \
    --gradient_accumulation_steps 8  \
    --gradient_checkpointing True \
    --evaluation_strategy "no" \
    --save_strategy "no" \
    --learning_rate 3e-4 \
    --warmup_ratio  0.03 \
    --lr_scheduler_type "cosine" \
    --max_grad_norm  0.3 \
    --logging_steps 1 \
    --do_train \
    --do_eval \
    --use_habana \
    --use_peft True \
    --pipelining_fwd_bwd True\
    --use_lazy_mode \
    --throughput_warmup_steps 3 \
    --lora_r 4 \
    --lora_alpha=16 \
    --lora_dropout=0.05 \
    --lora_target_modules "q_proj" "v_proj" "k_proj" "o_proj" \
    --max_seq_length 2048 \
    --adam_epsilon 3e-4 --use_flash_attention True --deepspeed /root/litang/QWEN/qwen2_finetune_bench_scripts/ds_config.json 2>&1 | tee log8x.txt
```

**Comment by ssarkar2** [comment]
```
time python3 sft.py  --model_name_or_path /root/sasarkar/qwen2_7b/ --dataset_name "philschmid/dolly-15k-oai-style"  --streaming False  --bf16 True  --subset ''  --output_dir ./model_qwen  --num_train_epochs 1  --per_device_train_batch_size 4  --evaluation_strategy "no"  --save_strategy "no"  --learning_rate 3e-4  --warmup_ratio  0.03  --lr_scheduler_type "cosine"  --max_grad_norm  0.3  --logging_steps 1  --do_train  --do_eval  --use_habana  --use_lazy_mode  --throughput_warmup_steps 3  --lora_r 4  --lora_alpha=16  --lora_dropout=0.05  --lora_target_modules "q_proj" "v_proj" "k_proj" "o_proj"  --max_seq_length 512  --adam_epsilon 1e-08 --packing False  --gradient_checkpointing True --pipelining_fwd_bwd True --deepspeed ds_config.json --max_steps 50
```

WIP

with --num_buckets 8:
TBD

File: tests/test_sft_train.py
Comment (regisss) [inline tests/test_sft_train.py:1]: We already have some tests for `sft.py` in https://github.com/huggingface/optimum-habana/blob/main/tests/test_examples.py. Is it possible to integrate this one there?

File: optimum/habana/trl/trainer/sft_trainer.py [hunk 1]
```diff
@@ -43,6 +46,49 @@
...
+                examples,
+                return_tensors="pt",
+                pad_to_multiple_of=bucketed_len,  # self.pad_to_multiple_of
+            )
+        else:
+            assert False, "This path has not been implemented/tested yet"
+            # https://github.com/huggingface/transformers/blob/v4.40.0/src/transformers/data/data_collator.py#L428
+            # batch = {
+            #    "input_ids": _torch_collate_batch(examples, self.tokenizer, pad_to_multiple_of=self.pad_to_multiple_of)
+            # }
```
Comment (regisss) [inline optimum/habana/trl/trainer/sft_trainer.py:76]: To remove?

File: tests/test_sft_train.py
Comment (ssarkar2) [reply tests/test_sft_train.py:1]: some extra command line args that I have:

--subset
--max_steps ( to keep the test short)
--lora_target_modules "q_proj" "v_proj" "k_proj" "o_proj"
--lr_scheduler_type 
--use_peft 
 --packing False 
etc

Not sure how to fit these into test_examples.py

Comment (regisss) [reply tests/test_sft_train.py:1]: You should be able to add any specific extra arguments in the `extra_arguments` field of the json baseline files, here is an example for Qwen: https://github.com/huggingface/optimum-habana/blob/59d182dfa9a4c0ba068a1fd118b96b8ff0057c96/tests/baselines/Qwen2_7B.json#L49

Comment (regisss) [reply tests/test_sft_train.py:1]: Let me know if that's good enough, otherwise we'll just keep the current version.

**Comment by ssarkar2** [comment]
Trying with llama2-7b:

`time python3 sft.py  --model_name_or_path /mnt/weka/data/llama_inference/Llama-2-7b-hf/ --dataset_name "philschmid/dolly-15k-oai-style"  --streaming False  --bf16 True  --subset ''  --output_dir ./model_qwen  --num_train_epochs 1  --per_device_train_batch_size 4  --evaluation_strategy "no"  --save_strategy "no"  --learning_rate 3e-4  --warmup_ratio  0.03  --lr_scheduler_type "cosine"  --max_grad_norm  0.3  --logging_steps 1  --do_train  --do_eval  --use_habana  --use_lazy_mode  --throughput_warmup_steps 3  --lora_r 4  --lora_alpha=16  --lora_dropout=0.05  --lora_target_modules "q_proj" "v_proj" "k_proj" "o_proj"  --max_seq_length 512  --adam_epsilon 1e-08 --packing False --num_bucket 8 --max_steps 100`

WIP

File: tests/test_sft_train.py
Comment (ssarkar2) [reply tests/test_sft_train.py:1]: @regisss
Ok, I can almost make test_examples.py work I think.

I have a variation in the cmd line:
```
if "72" in model_name:
        command += [
            "--max_steps",
            "50",
            "--gradient_checkpointing",
            "True",
            "--pipelining_fwd_bwd",
            "True",
        ] 
    else:
        command += ["--max_steps", "100"]
```

And also, I have a certain deepspeed config file, which I use for qwen72b model, which I create as a tmp file and dump and use 
```
if "72" in model_name:
            command += ["--deepspeed", fp.name]
```

Any suggestions how these might be incorporated in test_examples.py? For the if-else in the cmd line based on 7b or 72b model, I could enter the qwen2-7 settings in Qwebn2_7B.json, and for 72b model I could create a new json Qwen2-72B.json ? However, MODELS_TO_TEST_MAPPING has entry: "qwen2": [("Qwen/Qwen2-7B", "Habana/qwen")],, so then I'd have to change that as well maybe?

just checking if you have some easy/obvious solution, else I'll stick to the test file I have now

Comment (regisss) [reply tests/test_sft_train.py:1]: For the DeepSpeed config, you can also store it [here](https://github.com/huggingface/optimum-habana/tree/main/tests/configs) and pass an argument like this one: https://github.com/huggingface/optimum-habana/blob/773859564b1afbb95595f1f4a397a939fd993589/tests/baselines/bloom_7b1.json#L17
But that's up to you, using a temporary file is fine too.

> For the if-else in the cmd line based on 7b or 72b model, I could enter the qwen2-7 settings in Qwebn2_7B.json, and for 72b model I could create a new json Qwen2-72B.json ?

Yes, that sounds good!

> However, MODELS_TO_TEST_MAPPING has entry: "qwen2": [("Qwen/Qwen2-7B", "Habana/qwen")],, so then I'd have to change that as well maybe?

Yep, you would probably need to change it to:
```
"qwen2": [("Qwen/Qwen2-7B", "Habana/qwen"), ("name_of_72B_checkpoint", "Habana/qwen")]
```

Then, we will also need to add a rule in [this big conditional block](https://github.com/huggingface/optimum-habana/blob/773859564b1afbb95595f1f4a397a939fd993589/tests/test_examples.py#L223) to make sure the Qwen-72B is executed only for your test, but I can help with that at the end if you don't see how to do it.

Comment (regisss) [reply tests/test_sft_train.py:1]: @ssarkar2 Do you think we can the test integrated into test_examples.py?

**Review by regisss — APPROVED** [review]

**Pull request merged**